File: packages/ts/tests/integration/binding/value-fidelity.test.ts
// Binding-tier tests for value fidelity at the JS<->core boundary (#465):
// the numeric contract (out-of-range ints/BigInts error, never lossy) and
// BigInt->INTEGER, plus real onFile-error message propagation.
import { mkdtemp, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { DirSQL } from "dirsql";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

let dir: string;

beforeEach(async () => {
  dir = await mkdtemp(join(tmpdir(), "dirsql-fidelity-"));
});

afterEach(async () => {
  await rm(dir, { recursive: true, force: true });
});

function mkdb(onFile: () => Array<Record<string, unknown>>): DirSQL {
  return new DirSQL({
    root: dir,
    tables: [{ ddl: "CREATE TABLE t (v)", glob: "*.json", onFile }],
  });
}

describe("DirSQL integer range (query results)", () => {
  beforeEach(async () => {
    await writeFile(join(dir, "marker.json"), "{}");
  });

  it("throws when an integer beyond MAX_SAFE_INTEGER is read back", async () => {
    // 2^53 + 1: a valid i64, but not a safe JS Number.
    const db = mkdb(() => [{ v: 9007199254740993n }]);
    await expect(db.query("SELECT v FROM t")).rejects.toThrow(/safe integer/);
  });

  it("round-trips an integer at MAX_SAFE_INTEGER", async () => {
    const db = mkdb(() => [{ v: 9007199254740991n }]);
    const rows = await db.query("SELECT v FROM t");
    expect(rows[0].v).toBe(9007199254740991);
  });
});

describe("DirSQL BigInt onFile", () => {
  beforeEach(async () => {
    await writeFile(join(dir, "marker.json"), "{}");
  });

  it("maps a small BigInt to INTEGER", async () => {
    const db = mkdb(() => [{ v: 42n }]);
    const rows = await db.query("SELECT v FROM t");
    expect(rows[0].v).toBe(42);
  });

  it("skips the file for a BigInt beyond i64 range", async () => {
    // Since dirsql#714 an unrepresentable value is the hook's mistake and
    // costs its own file; the build resolves with that file contributing
    // nothing. Which file was skipped is not reachable here yet (#715).
    const db = mkdb(() => [{ v: 2n ** 64n }]);
    await db.ready;
    expect(await db.query("SELECT v FROM t")).toEqual([]);
  });
});

describe("DirSQL onFile error message", () => {
  beforeEach(async () => {
    await writeFile(join(dir, "marker.json"), "{}");
  });

  it("skips the file whose onFile threw", async () => {
    // Since dirsql#714 a throwing hook no longer fails the scan. The message
    // is carried on the core's failure list, which this binding cannot reach
    // yet (#715) -- so all that is observable here is the absent row.
    const db = mkdb(() => {
      throw new Error("bad JSON in posts/a.json: boom");
    });
    await db.ready;
    expect(await db.query("SELECT v FROM t")).toEqual([]);
  });
});

describe("DirSQL Buffer -> BLOB", () => {
  const payload = Buffer.from([0x00, 0x01, 0x02, 0xff, 0xfe]);

  async function roundTrip(data: unknown): Promise<unknown> {
    await writeFile(join(dir, "marker.json"), "{}");
    const db = new DirSQL({
      root: dir,
      tables: [
        {
          ddl: "CREATE TABLE blobs (name TEXT, data BLOB)",
          glob: "*.json",
          onFile: () => [{ name: "bin", data }],
        },
      ],
    });
    const rows = await db.query("SELECT * FROM blobs");
    expect(rows).toHaveLength(1);
    expect(rows[0].name).toBe("bin");
    return rows[0].data;
  }

  it("round-trips a Buffer through a BLOB column", async () => {
    const data = await roundTrip(payload);
    expect(Buffer.isBuffer(data)).toBe(true);
    expect(Buffer.compare(data as Buffer, payload)).toBe(0);
  });

  it("maps a Uint8Array to a BLOB and returns it as a Buffer", async () => {
    const data = await roundTrip(new Uint8Array(payload));
    expect(Buffer.isBuffer(data)).toBe(true);
    expect(Buffer.compare(data as Buffer, payload)).toBe(0);
  });
});

File: packages/python/tests/integration/binding/value_fidelity_test.py
"""Binding-tier tests for value fidelity at the Python<->core boundary (#465).

Covers the numeric contract (out-of-i64 ints raise, no lossy Real/TEXT
fallback) and the bytes-vs-sequence discrimination in ``py_to_value``.
"""

import os

import pytest

from dirsql import DirSQL, Table


def _db(tmp_dir, on_file, ddl="CREATE TABLE t (v)"):
    with open(os.path.join(tmp_dir, "marker.json"), "w") as f:
        f.write("{}")
    return DirSQL(tmp_dir, tables=[Table(ddl=ddl, glob="*.json", on_file=on_file)])


def describe_integer_range():
    # Since dirsql#714 an unrepresentable value is the hook's mistake and costs
    # its own file rather than the scan, so these assert the row never lands
    # instead of asserting a raise. The core still refuses the value and still
    # names it -- but that message rides on `scan_failures`, which this binding
    # cannot reach yet (#715). Until it can, "the value is named" is untestable
    # from Python, so `it_names_the_offending_value` has no assertion left to
    # make and is folded into the two range cases.
    @pytest.mark.asyncio
    async def it_skips_the_file_on_int_exceeding_i64_max(tmp_dir):
        db = _db(tmp_dir, lambda path: [{"v": 2**63}])
        await db.ready()
        assert await db.query("SELECT v FROM t") == []

    @pytest.mark.asyncio
    async def it_skips_the_file_on_int_below_i64_min(tmp_dir):
        db = _db(tmp_dir, lambda path: [{"v": -(2**63) - 1}])
        await db.ready()
        assert await db.query("SELECT v FROM t") == []

    @pytest.mark.asyncio
    async def it_round_trips_i64_max(tmp_dir):
        db = _db(tmp_dir, lambda path: [{"v": 2**63 - 1}])
        await db.ready()
        rows = await db.query("SELECT v FROM t")
        assert rows[0]["v"] == 2**63 - 1


def describe_bytes_vs_sequence():
    @pytest.mark.asyncio
    async def it_maps_bytes_to_blob(tmp_dir):
        db = _db(tmp_dir, lambda path: [{"v": b"\x01\x02\x03"}])
        await db.ready()
        rows = await db.query("SELECT v FROM t")
        assert rows[0]["v"] == b"\x01\x02\x03"

    @pytest.mark.asyncio
    async def it_does_not_probe_a_small_int_list_as_bytes(tmp_dir):
        db = _db(tmp_dir, lambda path: [{"v": [1, 2, 3]}])
        await db.ready()
        rows = await db.query("SELECT v FROM t")
        assert isinstance(rows[0]["v"], str)

    @pytest.mark.asyncio
    async def it_treats_int_lists_the_same_regardless_of_magnitude(tmp_dir):
        for payload in ([1, 2, 3], [1, 2, 300]):
            db = _db(tmp_dir, lambda path, p=payload: [{"v": p}])
            await db.ready()
            v = (await db.query("SELECT v FROM t"))[0]["v"]
            assert isinstance(v, str)
